src/signature.rs

```diff
@@ -0,0 +1,202 @@
+
+            let signature = SecpRecoverableSignature::from_compact(self.as_ref(), v)
+                .unwrap_or_else(|_| {
+                    SecpRecoverableSignature::from_compact(&[0u8; 64], v)
+                        .expect("Zeroed signature is infallible")
+                });
+
+            signature
+        }
+
+        pub(crate) fn from_secp(signature: SecpRecoverableSignature) -> Self {
+            let (v, mut signature) = signature.serialize_compact();
+
+            let v = v.to_i32();
+
+            signature[32] |= (v << 7) as u8;
+
+            // Safety: the security of this call reflects the security of secp256k1 FFI
+            unsafe { Signature::from_bytes_unchecked(signature) }
+        }
+
+        /// Truncate the recovery id from the signature, producing a valid `secp256k1`
+        /// representation.
+        pub(crate) fn truncate_recovery_id(&mut self) {
+            self.as_mut()[32] &= 0x7f;
+        }
+
+        /// Sign a given message and compress the `v` to the signature
+        ///
+        /// The compression scheme is described in
+        /// <https://github.com/FuelLabs/fuel-specs/blob/master/specs/protocol/cryptographic_primitives.md#public-key-cryptography>
+        pub fn sign(secret: &SecretKey, message: &Message) -> Self {
+            let secret = secret.borrow();
+            let message = message.to_secp();
+
+            let signature = Secp256k1::signing_only().sign_recoverable(&message, secret);
+
+            Signature::from_secp(signature)
+        }
+
+        /// Recover the public key from a signature performed with
+        /// [`Signature::sign`]
+        ///
+        /// It takes the signature as owned because this operation is not idempotent. The taken
+        /// signature will not be recoverable. Signatures are meant to be single use, so this
+        /// avoids unnecessary copy.
+        pub fn recover(mut self, message: &Message) -> Result<PublicKey, Error> {
+            let signature = self.to_secp();
+            let message = message.to_secp();
+
+            let pk = Secp256k1::new()
+                .recover(&message, &signature)
+                .map(|pk| PublicKey::from_secp(&pk))?;
+
+            Ok(pk)
+        }
+
+        /// Verify a signature produced by [`Signature::sign`]
+        ///
+        /// It takes the signature as owned because this operation is not idempotent. The taken
+        /// signature will not be recoverable. Signatures are meant to be single use, so this
+        /// avoids unnecessary copy.
+        pub fn verify(mut self, pk: &PublicKey, message: &Message) -> Result<(), Error> {
+            self.truncate_recovery_id();
+
+            let signature = Secp256k1Signature::from_compact(self.as_ref())?;
+
+            let message = message.to_secp();
+            let pk = pk.to_secp()?;
+
+            Secp256k1::verification_only().verify(&message, &signature, &pk)?;
+
+            Ok(())
```

Comment:
Is there a reason you don't call `recover`, then compare the pubkeys? Seems logically easier to grok.

Reply:
The `verify` *should be* (need to check) cheaper than `recover`. That would be the only reason.

Reply:
My personal preference would be to implement it as I describe above since it's simpler and less code repeating, then file a good first issue to benchmark and potentially optimize. But won't veto.

Reply:
I did a preliminary check, and `verify` is `~10%` cheaper than `recover`

Reply:
https://github.com/FuelLabs/fuel-crypto/pull/2/commits/a3218907e960770e9e142c1934769ade3bbb9279